### PR TITLE
Fix nan handling in output buffer for CGemv

### LINKED ISSUE #3887: Nans occuring in function unless compiled with DebugMode, NanGuardMode or Print
While implementing L-BFGS in theano, I came across a really strange bug: My two loop recursion results in a nan vector. However, if I print intermediate values with `theano.printing.Print` (see comments in code below) or if I compile the function with `DebugMode` or `NanGuardMode`, the nans go away. At some specific places, `Print`-Operators show actual nans (see comments below). Unfortunately, I was not able to reduce the example further than in the code below. However, I noticed that the behaviour was not present in 7.0 but is present in current master. Running git bisect, I was able to pinpoint the bug to commit 599995d062437d20052fd649cc029a61dbc82f5f, which consists of only one line.

```python
import numpy as np
import theano
import theano.tensor as T
from theano.ifelse import ifelse

def Print(name, variable):
    return theano.printing.Print(name)(variable)

def scalar_product(vs, ws):
    """Calculate scalar product over to lists of parameters"""
    assert len(vs) == len(ws)
    #return T.sum([Print('v', v).dot(Print('w', w)) for v, w in zip(vs, ws)])
    return T.sum([v.dot(w) for v, w in zip(vs, ws)])

def two_loop_recursion(gradients, Ys, Ss, k, apply_Hk0):

    n_params = len(gradients)

    not_first_step = T.gt(k,0)

    #### First loop

    def f(i, *args):
        qs = args[:n_params]
        alpha = args[n_params]
        Ys = args[n_params+1:2*n_params+1]
        Ss = args[2*n_params+1:]
        assert len(Ss) == n_params

        valid_step = T.ge(i, 0)

        store_index = i % Ys[0].shape[0]

        y_is = []
        s_is = []
        for Y, S in zip(Ys, Ss):
            y_is.append(Y[store_index])
            s_is.append(S[store_index])

        ydots = scalar_product(y_is, s_is)
        rho = 1/ydots

        sdotq = scalar_product(s_is, qs)
        alpha = rho * sdotq

        new_qs = []

        for q, y_i in zip(qs, y_is):
            new_q = q - alpha*y_i 
            new_q = ifelse(valid_step, new_q, q)
            new_qs.append(new_q)

        new_alpha = ifelse(valid_step, alpha, np.array(0.0, dtype=theano.config.floatX))

        return tuple(new_qs + [new_alpha])

    qs = gradients
    alpha = T.zeros(())

    loop_indices = T.arange(k-Ys[0].shape[0], k)


    scan_results, updates = theano.scan(f, sequences=[loop_indices],
                                       outputs_info=qs+[alpha],
                                       non_sequences=Ys+Ss,
                                       strict=True,
                                       go_backwards=True)

    qs_results = scan_results[:n_params]

    alpha_result = scan_results[n_params]

    alpha = alpha_result[::-1]

    rs = apply_Hk0(qs, Ss, Ys, k)

    # This makes the nans go away
    #rs = [Print('r', r) for r in rs]

    #### Second loop

    def f2(i, alpha_i, *args):

        rs = args[:n_params]

        rs = [r for r in rs]

        Ys = args[n_params:2*n_params]
        Ss = args[2*n_params:]
        assert len(Ss) == n_params

        valid_step = T.ge(i, 0)

        store_index = i % Ys[0].shape[0]

        y_is = []
        s_is = []
        for Y, S in zip(Ys, Ss):
            y_is.append(Y[store_index])
            s_is.append(S[store_index])

        ydots = scalar_product(y_is, s_is)
        rho = 1/ydots
        # printing rho and ydotr makes rho not nan and ydotr nan
        #rho = Print('rho', rho)

        ydotr = scalar_product(y_is, rs)

        # ydotr is not nan if rho is not printed, otherwise it is nan
        ydotr = Print('ydotr', ydotr)
        beta = rho * ydotr

        #beta is always printed as nan, as long rs was not printed outside of the scan
        beta = Print('beta', beta)


        new_rs = []
        for r, s_i in zip(rs, s_is):
            new_r = r + s_i*(alpha_i - beta)
            new_r = ifelse(valid_step, new_r, r)
            new_rs.append(new_r)

        return tuple(new_rs)

    results, updates = theano.scan(f2, sequences=[loop_indices, alpha],
                             outputs_info=tuple(rs),
                             non_sequences=Ys + Ss,
                             go_backwards=False,
                             strict=True)
    if n_params == 1:
        # In this case, theano returns only one object instead of a tuple
        rs = [results[-1]]
    else:
        rs = [result[-1] for result in results]

    return rs

def apply_Hk0(qs, Ss, Ys, k):
    """
    H_k^0 q := gamma_k q with

    gamma_k = s_{k-1}^Ty_{k_1} / y_{k-1}^Ty_{k-1}  (7.20)

    """
    ss = [S[(k - 1) % S.shape[0]] for S in Ss]
    ys = [Y[(k - 1) % Y.shape[0]] for Y in Ys]

    # any of this lines makes the nans go away
    #ss = [Print('s', s) for s in ss]
    #ys = [Print('y', y) for y in ys]

    gamma_k = scalar_product(ss, ys) / scalar_product(ys, ys)

    gamma_k = ifelse(T.gt(k,0), gamma_k, np.array(1.0, dtype=theano.config.floatX))
    return [gamma_k * q for q in qs]

# Setting up example

N = 3
m = 20
x_0 = -np.ones(N)
x = theano.shared(x_0)
y = 0.2*(T.abs_(x)**2.3).sum()
params = [x]
gradients = T.grad(y, params)

k = theano.shared(0)
Ss = [theano.shared(np.random.standard_normal((m, len(_x.get_value()))), name='S') for _x in params]
Ys = [theano.shared(np.random.standard_normal((m, len(_x.get_value()))), name='Y') for _x in params]

rr = two_loop_recursion(gradients, Ys, Ss, k, apply_Hk0)
f_theano = theano.function([], rr)

k.set_value(1)
assert not np.any(np.isnan(f_theano()))
```

### PR DESCRIPTION
**Pull request by abergeron**
This fixes #3887

### CODE REVIEW

**Comment by abergeron** [comment]
I've fixed the failing tests.

File: theano/tensor/blas_c.py [hunk 1]
```diff
@@ -422,174 +386,114 @@ def gemv_c_code(aa, xx, yy, zz, alpha, beta, destructive, fail,
     // copy aa if not destructive
```
Comment (lamblin) [inline theano/tensor/blas_c.py]: "aa" is now named "y", update the comment

File: theano/tensor/blas_c.py [hunk 2]
```diff
@@ -666,94 +580,88 @@ def gemv_c_code(aa, xx, yy, zz, alpha, beta, destructive, fail,
             // the matrix has at least one dim of length 0
             // so we do this loop, which either iterates over 0 elements
             // or else it does the right thing for length-0 x.
```
Comment (lamblin) [inline theano/tensor/blas_c.py]: x -> A?
There is also an "xx" in an error message above.

**Comment by lamblin** [comment]
Otherwise, it looks good to me.
It has the added benefit of not storing the behaviour when pickling a CGemv, but to re-detect it when actually needing to generate the C code.

**Comment by abergeron** [comment]
I fixed the names.

**Pull request merged**